### PR TITLE
Let `LocalFileLinter` reuse the `Dependency` instead falling back on its path

### PR DESCRIPTION
**Pull request by JCZuurmond**
## Changes

Let `LocalFileLinter` reuse the `Dependency` instead falling back on its path as we introduce duplicate ways of loading files. 

In addition to that, the PR contains a clean up of:
- Moving non-linting file classes from `source_code/linters/files.py` to `source_code/files.py` to avoid circular imports
- Extend `MockPathLookup` to verify which paths are resolved
- Make `current_session` public on `LinterContext` for reuse instead of passing it next to the `LinterContext`
- Improve testing and handling corner cases for `FileLoader`

### Linked issues

Progresses #3514
Breaks up #3520

### Functionality

- [x] modified existing command: `databricks labs ucx migrate-local-code`

### Tests

- [ ] manually tested
- [x] modified and added unit tests
- [x] modified and added integration tests

### CODE REVIEW

File: src/databricks/labs/ucx/source_code/files.py [hunk 1]
```diff
@@ -0,0 +1,98 @@
...
+        if self.language == Language.PYTHON:
+            context = parent.new_dependency_graph_context()
+            analyzer = PythonCodeAnalyzer(context, self._original_code)
+            problems = analyzer.build_graph()
+            for idx, problem in enumerate(problems):
+                if problem.has_missing_path():
+                    problems[idx] = dataclasses.replace(problem, source_path=self._path)
+            return problems
+        # supported language that does not generate dependencies
+        if self.language == Language.SQL:
```
Comment (pritishpai) [inline src/databricks/labs/ucx/source_code/files.py:47]: Can we combine this and the other return as SQL is also unsupported and probably doesn't need an if statement

File: src/databricks/labs/ucx/source_code/linters/context.py [hunk 2]
```diff
@@ -72,14 +74,14 @@ def __init__(
...
         python_linters += [
-            DirectFsAccessPyLinter(session_state),
-            DBRv8d0PyLinter(dbr_version=session_state.dbr_version),
-            SparkConnectPyLinter(session_state),
-            DbutilsPyLinter(session_state),
+            DirectFsAccessPyLinter(self.session_state),
+            DBRv8d0PyLinter(dbr_version=self.session_state.dbr_version),
+            SparkConnectPyLinter(self.session_state),
+            DbutilsPyLinter(self.session_state),
             DirectFsAccessSqlPylinter(sql_direct_fs),
```
Comment (pritishpai) [inline src/databricks/labs/ucx/source_code/linters/context.py:81]: Should the sql_direct_fs also be a part of LinterContext object?

**Review by pritishpai — APPROVED** [review]
Commented a few nits. Approving so that you can merge in the AM.

File: src/databricks/labs/ucx/source_code/linters/context.py (hunk 2 above)
Comment (JCZuurmond) [reply src/databricks/labs/ucx/source_code/linters/context.py:81]: All linters should be part of the linter context, see docstring added in https://github.com/databrickslabs/ucx/pull/3638/files

File: src/databricks/labs/ucx/source_code/files.py (hunk 1 above)
Comment (JCZuurmond) [reply src/databricks/labs/ucx/source_code/files.py:47]: I prefer to keep it like this because SQL is supported, it just does not return a dependency graph, see #3639

**Comment by JCZuurmond** [comment]
Surpassed by #3641
